ci: Disable macOS functional tests on forked repos to avoid timeouts

hebasto opened the pull request:
See: https://github.com/bitcoin-core/gui/issues/5#issuecomment-656819184

Additionally, this PR:
- updates macOS image to the recent 10.15.5 version
- drops Homebrew caching as the Travis Homebrew addon have been used since #18438

My forked repo build: https://travis-ci.org/github/hebasto/bitcoin/jobs/707200431

promag commented:
I think commit "ci: Update macOS image to 10.15.5 on Travis" doesn't belong here. What is the motivation? ACK remaining commits.

hebasto commented:
@promag 
> I think commit "ci: Update macOS image to 10.15.5 on Travis" doesn't belong here. What is the motivation? ACK remaining commits.

The mentioned commit has been dropped.

promag commented:
ACK 0b9e1e614c3b4fa747d6c1dbeedef71ce834fe94

MarcoFalke commented on .travis.yml:
```diff
@@ -43,11 +43,12 @@ before_script:
   - if [ $TRAVIS_REPO_SLUG = "bitcoin/bitcoin" ]; then export CONTINUE=1; fi  # continue on repos with extended build time (90 minutes)
   - if [ $CONTINUE = "1" ]; then set -o errexit; source ./ci/test/06_script_a.sh; else set +o errexit; echo "$CACHE_ERR_MSG"; false; fi
-  - if [ $SECONDS -gt 2000 ]; then export CONTINUE=0; fi  # Likely the build took very long; The tests take about 1000s, so we should abort if we have less than 50*60-1000=2000s left
+  - if [ $SECONDS -gt 2000 && $LONG_RUNNING_TEST = "1" ]; then export CONTINUE=0; fi  # Likely the build took very long; The tests take about 1000s, so we should abort if we have less than 50*60-1000=2000s left
```
Should this be an `&&` or `||` and why? Also, why use the runtime heuristic in combination with the heuristic in line 46. Wouldn't it be easier to simply export `LONG_RUNNING_TEST` for the configs that need it? I.e. right in the line where the other env vars are defined:

```
      env: >-
        FILE_ENV="..."
        LONG_...

hebasto replied:
^ this will not work as the `LONG_RUNNING_TEST` actually should depends on the `TRAVIS_REPO_SLUG`.

Another suggestion is to `export LONG_RUNNING_TEST="true"` by default in `00_setup_env.sh` and set it in a job setup script `00_setup_env_*.sh`:
```
if [ "$TRAVIS_REPO_SLUG" != "bitcoin/bitcoin" ]; then
  export RUN_FUNCTIONAL_TESTS="false"
  export LONG_RUNNING_TEST="false"
fi
```

MarcoFalke replied:
If you don't set `LONG_RUNNING_TEST` to anything for the macos config, everything will work fine in all repos, no?

* `bitcoin/bitcoin` will run the functional tests, and forked repos won't. Independent of `LONG_RUNNING_TEST`

hebasto replied:
> If you don't set `LONG_RUNNING_TEST` to anything for the macos config, everything will work fine in all repos, no?
> 
>  - `bitcoin/bitcoin` will run the functional tests, and forked repos won't. Independent of `LONG_RUNNING_TEST`

With cleared (or failed) cache forked repos could experience timeouts.

The mean of `LONG_RUNNING_TEST` is:
- it is set -- do not continue tests if not enough time remains. This should be done to avoid timeouts in forked repos
- it is unset -- do not take into account the remaining time, just continue. This should be done in cases of manual disabled tests in forked repos.

Actually, the `LONG_RUNNING_TEST` only affects forked repos due to the following code:
https://github.com/bitcoin/bitcoin/blob/4db44acf2d5d3b40943c94b3960f42838255b7ad/.travis.yml#L51-L52

MarcoFalke replied:
Ok, so this seems to affect more than just macos. Hmm. I am not sure if it is expected to fork the repo, run the tests on a commit and then see a green checkmark when the functional tests haven't even been run. Then the commit is proposed to bitcoin/bitcoin and the tests fail.

Excluding macOS is fine because I am not aware this ever detected a bug that was not detected by any of the other builds. Though, disabling all tests for forked repos seems a bit too much, no?

If travis can't give us what we need, we should switch providers instead of disabling tests conditionally.

MarcoFalke commented on .travis.yml:
```diff
@@ -43,11 +41,12 @@ before_script:
   - for i in {1..4}; do echo "$(sleep 500)" ; done &
   - set -o errexit; source ./ci/test/05_before_script.sh &> "/dev/null"
 script:
+  - if [[ -n "$USE_VALGRIND" || "$RUN_FUNCTIONAL_TESTS" = "true" || "$RUN_FUZZ_TESTS" = "true" ]]; then export LONG_RUNNING_TEST=1; fi
```
are the fuzz tests timing out as well these days?

practicalswift replied:
I hope not! Do we have any evidence suggesting that? If so I'll investigate how we could speed things up.

maflcko reviewed (approved):
ACK

maflcko reviewed (commented):
withdraw ACK for now

maflcko reviewed (changes requested):
withdraw ACK for now

hebasto commented:
Updated 0b9e1e614c3b4fa747d6c1dbeedef71ce834fe94 -> 12f642de257a0280e014ad8660f54ee45574a637 ([pr19495.02](https://github.com/hebasto/bitcoin/commits/pr19495.02) -> [pr19495.03](https://github.com/hebasto/bitcoin/commits/pr19495.03), [diff](https://github.com/hebasto/bitcoin/compare/pr19495.02..pr19495.03)):

- timeout changes affect only native macOS build now
- added _"ci: Fix configure options for macOS builds"_ commit

--

However, still working on performance improving for the native macOS build on Travis.

MarcoFalke commented on .travis.yml:
```diff
@@ -48,7 +46,7 @@ script:
   - if [ $CONTINUE = "1" ]; then set -o errexit; source ./ci/test/06_script_a.sh; else set +o errexit; echo "$CACHE_ERR_MSG"; false; fi
   - if [ $SECONDS -gt 2000 ]; then export CONTINUE=0; fi  # Likely the build took very long; The tests take about 1000s, so we should abort if we have less than 50*60-1000=2000s left
-  - if [ $TRAVIS_REPO_SLUG = "bitcoin/bitcoin" ]; then export CONTINUE=1; fi  # continue on repos with extended build time (90 minutes)
+  - if [ $TRAVIS_REPO_SLUG = "bitcoin/bitcoin" || $CONTINUE_IN_FORKED_REPO = "true" ]; then export CONTINUE=1; fi  # continue on repos with extended build time (90 minutes)
```
```suggestion
  - if [ $TRAVIS_REPO_SLUG = "bitcoin/bitcoin" ]; then export CONTINUE=1; fi  # continue on repos with extended build time (90 minutes)
```

Is this needed? With a warm cache, this should already continue, and without a warm cache it might or might not continue.

I think we should either keep track of when to continue manually for all configs or with the time-heuristic for all configs, not a mix of both.

maflcko reviewed (commented):
ACK

hebasto committed:
ci: Disable functional tests on forked repos to avoid timeouts for macOS

hebasto committed:
ci: Do not activate Travis ccache caching strategy

It is sufficient to specify CCACHE_DIR to cache.
Also this change fixes ccache on native macOS build.

hebasto committed:
ci: Drop Homebrew caching while using Homebrew addon on Travis

hebasto committed:
ci: Fix configure options for macOS builds

hebasto commented:
Updated 12f642de257a0280e014ad8660f54ee45574a637 -> 60824b3c3a4221a31c3638008b05fecc040c3df2 ([pr19495.03](https://github.com/hebasto/bitcoin/commits/pr19495.03) -> [pr19495.04](https://github.com/hebasto/bitcoin/commits/pr19495.04)):

- addressed @MarcoFalke's [comment](https://github.com/bitcoin/bitcoin/pull/19495#discussion_r453527250)
> I think we should either keep track of when to continue manually for all configs or with the time-heuristic for all configs, not a mix of both.

The time-heuristic is used for all configs. It could customized per build with the `EXPECTED_TESTS_DURATION_IN_SECONDS` environment variable.

- added _"ci: Do not activate Travis ccache caching strategy"_ commit that fixes ccache for native macOS build on Travis
